### ISSUE TITLE
Is there any mechanism to protect the charts that I developed?

### ISSUE THREAD

**Issue by jeremyxu2010** [helm/helm#5043]
The scenario:
My company spent a certain amount of cost to develop the corresponding charts for the partner's applications. These charts greatly accelerated the deployment efficiency of the partner's applications. But these charts are placed directly on the charts repo currently, my competitors can download these "charts" easily, and put them into their products. Since charts are already a kind of codes, is there any mechanism to protect the "charts" I developed?

**Comment by maorfr** [comment]
What is your chart repo?

**Comment by maorfr** [comment]
What do you mean by "the chart repository in our products"?
Can you provide mode details?

And in general, can you enable authentication?

And to be even more general, even if your competition use your charts, do they have access to the docker images? Are those protected? If so, the charts will not help them, it is just a package.

**Comment by jeremyxu2010** [comment]
> What do you mean by "the chart repository in our products"?
> Can you provide mode details?
> 
> And in general, can you enable authentication?
> 
> And to be even more general, even if your competition use your charts, do they have access to the docker images? Are those protected? If so, the charts will not help them, it is just a package.

Our product is a software system which managed multiple k8s clusters. And we use helm to distribute our partner's  software to k8s clusters. In our software system there is a chart repository which contains the charts which is developed by us. So our competitors can deploy our products, and get the charts from the chart repository.

We and our competitors can all get the docker images, because these images are provided by our partners. Compared to competitors, our advantage is having developed some charts, which greatly speed up the efficiency of software deployment. We don't want to give up this advantage.

**Comment by maorfr** [comment]
Can you enable authentication on your chart repository?

**Comment by carlosjgp** [comment]
@jeremyxu2010 you can use a private chart repository (Nexus, S3 bucket,...) and only whitelist the IPs you want to access it

Helm already provides few mechanisms to protect your Chart repositories like certificates or basic authentication
https://docs.helm.sh/helm/#helm-repo-add

**Comment by jeremyxu2010** [comment]
> @jeremyxu2010 you can use a private chart repository (Nexus, S3 bucket,...) and only whitelist the IPs you want to access it
> 
> Helm already provides few mechanisms to protect your Chart repositories like certificates or basic authentication
> https://docs.helm.sh/helm/#helm-repo-add

I know that the chart repository itself has HTTPS client authentication and HTTP basic authentication. But since chart itself is not encrypted, a legitimate user can get all charts and then take these charts away.

**Comment by maorfr** [comment]
If I had to define the problem as I understand it, it would be:
An experience user has access to all charts.

The solution should be to limit an experience user access.

There are many ways to implement this. One of them can be to have 2 chart repos. One for experience users, one for a regular user.

But even then, your competition can pay for a single regular user, and get all your charts.

Even if you use encryption, authentication, authorization, or anything else,.your competition can get your charts.

To be honest, in my opinion, this is a bit of a strange disscussion to have in the context of open source software.

I am sure that charts is not what makes you stand out.

**Comment by bacongobbler** [comment]
I disagree here. If the user wants to keep their charts encrypted, then they should be allowed to do so, provided the complexity introduced makes sense. It's not our place to say whether or not their value proposition is their Helm charts (or any other part of their product for that matter).

We just closed #5048 because it relies on Tiller which has been removed in Helm 3. However, if you want to try again and write another PR against Helm 3, that'd be great. We'd also appreciate supporting documentation to go with the feature so others know how to use this new functionality. Thanks!
